Mathlib/CategoryTheory/Sites/IsSheafOneHypercover.lean

```diff
@@ -0,0 +1,167 @@
+`GrothendieckTopology.IsGeneratedByOneHypercovers.{max u v} J` holds (TODO).
+
+## TODO
+* Show that functors which preserve 1-hypercovers are continuous.
+* Refactor `DenseSubsite` using `1`-hypercovers.
+
+-/
+
+universe w v v' u u'
+
+namespace CategoryTheory
+
+open Category Limits
+
+variable {C : Type u} [Category.{v} C] (J : GrothendieckTopology C)
+  {A : Type u'} [Category.{v'} A]
+
+namespace GrothendieckTopology
+
+/-- A family of 1-hypercovers consists of the data of a predicate on
+`OneHypercover.{w} J X` for all `X`. -/
+abbrev OneHypercoverFamily := ∀ ⦃X : C⦄, OneHypercover.{w} J X → Prop
+
+namespace OneHypercoverFamily
+
+variable {J}
+variable (H : OneHypercoverFamily.{w} J)
+
+/-- A family of 1-hypercovers generates the topology if any covering sieve
+contains the sieve generated by the underlying covering of one of these 1-hypercovers.
+See `OneHypercoverFamily.isSheaf_iff` for the characterization of sheaves. -/
+class IsGenerating : Prop where
+  le {X : C} (S : Sieve X) (hS : S ∈ J X) :
+    ∃ (E : J.OneHypercover X) (_ : H E), E.sieve₀ ≤ S
+
+variable [H.IsGenerating]
+
+lemma exists_oneHypercover {X : C} (S : Sieve X) (hS : S ∈ J X) :
+    ∃ (E : J.OneHypercover X) (_ : H E), E.sieve₀ ≤ S :=
+  IsGenerating.le _ hS
+
+variable (P : Cᵒᵖ ⥤ A)
+
+namespace IsSheafIff
+
+variable (hP : ∀ ⦃X : C⦄ (E : J.OneHypercover X) (_ : H E), Nonempty (IsLimit (E.multifork P)))
+
+lemma hom_ext {X : C} (S : Sieve X) (hS : S ∈ J X) {T : A}
+    {x y : T ⟶ P.obj (Opposite.op X)}
+    (h : ∀ ⦃Y : C⦄ (f : Y ⟶ X) (_ : S f), x ≫ P.map f.op = y ≫ P.map f.op) :
+    x = y := by
+  obtain ⟨E, hE, le⟩ := H.exists_oneHypercover S hS
+  exact Multifork.IsLimit.hom_ext (hP E hE).some (fun j => h _ (le _ (Sieve.ofArrows_mk _ _ _)))
+
+variable {P H}
+variable {X : C} {S : Sieve X} {E : J.OneHypercover X} (hE : H E) (le : E.sieve₀ ≤ S)
+
+section
+
+variable (F : Multifork (Cover.index ⟨S, J.superset_covering le E.mem₀⟩ P))
+
+/-- Auxiliary definition for `isLimit`. -/
+noncomputable def lift : F.pt ⟶ P.obj (Opposite.op X) :=
+  Multifork.IsLimit.lift (hP E hE).some
+    (fun i => F.ι ⟨_, E.f i, le _ (Sieve.ofArrows_mk _ _ _)⟩)
+    (fun ⟨⟨i₁, i₂⟩, j⟩ => F.condition
+        { h₁ := le _ ((Sieve.ofArrows_mk _ _ i₁))
+          h₂ := le _ ((Sieve.ofArrows_mk _ _ i₂))
+          w := E.w j })
+
+@[reassoc]
+lemma fac' (i : E.I₀) :
+    lift hP hE le F ≫ P.map (E.f i).op =
+      F.ι ⟨_, E.f i, le _ (Sieve.ofArrows_mk _ _ _)⟩ :=
+  Multifork.IsLimit.fac (hP E hE).some _ _ i
+
+lemma fac {Y : C} (f : Y ⟶ X) (hf : S f) :
+    lift hP hE le F ≫ P.map f.op = F.ι ⟨Y, f, hf⟩ := by
+  apply hom_ext H P hP _ (J.pullback_stable f E.mem₀)
+  intro Z g
+  rintro ⟨T, a, b, hb, fac⟩
+  obtain ⟨i, rfl, hi⟩ := hb.exists
+  dsimp at hi
+  rw [id_comp] at hi
+  subst hi
+  rw [assoc, ← P.map_comp, ← op_comp, ← fac,
+    op_comp, P.map_comp, fac'_assoc]
+  exact F.condition
+    { h₁ := le _ (Sieve.ofArrows_mk _ _ _)
+      h₂ := hf
+      w := fac }
```

Comment:
```suggestion
  cases hb with
  | mk i => 
    rw [assoc, ← P.map_comp, ← op_comp, ← fac,
      op_comp, P.map_comp, fac'_assoc]
    exact F.condition
      { h₁ := le _ (Sieve.ofArrows_mk _ _ _)
        h₂ := hf
        w := fac }
```

Reply:
Usually, I would use `obtain ⟨i⟩ := hb`, but this occasionnaly make certain variable names unreachable (here `Y` becomes `Y✝`), which was the reason for the `ofArrows.exists` lemma. I am happy to see that `cases` does not have this problem. Here, I have used the syntax `cases' hb with i`.